File: app/src/main/java/com/cwlarson/deviceid/database/Software.java
package com.cwlarson.deviceid.database;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.Context;
import android.content.pm.ConfigurationInfo;
import android.os.Build;
import android.text.format.DateFormat;
import android.util.Log;

import com.cwlarson.deviceid.databinding.Item;
import com.cwlarson.deviceid.databinding.ItemType;
import com.cwlarson.deviceid.databinding.UnavailableItem;
import com.cwlarson.deviceid.databinding.UnavailableType;
import com.cwlarson.deviceid.util.SystemProperty;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

class Software {
    private final String TAG = Software.class.getSimpleName();
    private final Context context;

    Software(Activity activity, AppDatabase db){
        this.context = activity.getApplicationContext();
        //Set Software Tiles
        ItemAdder itemAdder = new ItemAdder(context, db);
        itemAdder.addItems(getAndroidVersion());
        itemAdder.addItems(getPatchLevel());
        itemAdder.addItems(getPreviewSDKInt());
        itemAdder.addItems(getDeviceBuildVersion());
        itemAdder.addItems(getBuildBaseband());
        itemAdder.addItems(getBuildKernel());
        itemAdder.addItems(getBuildDate());
        itemAdder.addItems(getBuildNumber());
        itemAdder.addItems(getBuildBoard());
        itemAdder.addItems(getBuildBootloader());
        itemAdder.addItems(getBuildBrand());
        itemAdder.addItems(getBuildDevice());
        itemAdder.addItems(getBuildDisplay());
        itemAdder.addItems(getBuildFingerprint());
        itemAdder.addItems(getBuildHardware());
        itemAdder.addItems(getBuildHost());
        itemAdder.addItems(getBuildTags());
        itemAdder.addItems(getBuildType());
        itemAdder.addItems(getBuildUser());
        itemAdder.addItems(getOpenGLVersion());
        itemAdder.addItems(getGooglePlayServicesVersion());
        itemAdder.addItems(getGooglePlayServicesInstallDate());
        itemAdder.addItems(getGooglePlayServicesUpdatedDate());
    }

    private enum Codenames {
        BASE, BASE_1_1,
        CUPCAKE,
        CUR_DEVELOPMENT,
        DONUT,
        ECLAIR, ECLAIR_MR1, ECLAIR_MR2,
        FROYO,
        GINGERBREAD, GINGERBREAD_MR1,
        HONEYCOMB, HONEYCOMB_MR1, HONEYCOMB_MR2,
        ICE_CREAM_SANDWICH, ICE_CREAM_SANDWICH_MR1,
        JELLY_BEAN, JELLY_BEAN_MR1, JELLY_BEAN_MR2,
        KITKAT, KITKAT_WATCH,
        LOLLIPOP, LOLLIPOP_MR1,
        MARSHMALLOW,
        NOUGAT, NOUGAT_MR1,
        OREO, OREO_MR1;

        public static Codenames getCodename()
        {
            int api = Build.VERSION.SDK_INT;
            switch (api) {
                case 1:
                    return BASE;
                case 2:
                    return BASE_1_1;
                case 3:
                    return CUPCAKE;
                case 4:
                    return DONUT;
                case 5:
                    return ECLAIR;
                case 6:
                    return ECLAIR_MR1;
                case 7:
                    return ECLAIR_MR2;
                case 8:
                    return FROYO;
                case 9:
                    return GINGERBREAD;
                case 10:
                    return GINGERBREAD_MR1;
                case 11:
                    return HONEYCOMB;
                case 12:
                    return HONEYCOMB_MR1;
                case 13:
                    return HONEYCOMB_MR2;
                case 14:
                    return ICE_CREAM_SANDWICH;
                case 15:
                    return ICE_CREAM_SANDWICH_MR1;
                case 16:
                    return JELLY_BEAN;
                case 17:
                    return JELLY_BEAN_MR1;
                case 18:
                    return JELLY_BEAN_MR2;
                case 19:
                    return KITKAT;
                case 20:
                    return KITKAT_WATCH;
                case 21:
                    return LOLLIPOP;
                case 22:
                    return LOLLIPOP_MR1;
                case 23:
                    return MARSHMALLOW;
                case 24:
                    return NOUGAT;
                case 25:
                    return NOUGAT_MR1;
                case 26:
                    return OREO;
                case 27:
                    return OREO_MR1;
                case 1000:
                    return CUR_DEVELOPMENT;
                default:
                    return null;
            }
        }
    }

    private Item getAndroidVersion() {
        Item item = new Item("Android Version", ItemType.SOFTWARE);
        try {
            //noinspection ConstantConditions
            String versionName = (Codenames.getCodename() == null) ? "" : Codenames.getCodename().toString();
            item.setSubtitle(Build.VERSION.RELEASE+" ("+String.valueOf(Build.VERSION.SDK_INT)+") " +
                versionName);
        } catch (Exception e) {
            Log.w(TAG, "Null in getAndroidVersion");
        }
        return item;
    }

    private Item getPatchLevel() {
        Item item = new Item("Security Patch Level", ItemType.SOFTWARE);
        try {
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                try {
                    SimpleDateFormat template = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
                    Date patchDate = template.parse(Build.VERSION.SECURITY_PATCH);
                    String format = DateFormat.getBestDateTimePattern(Locale.getDefault(), "dMMMMyyyy");
                    item.setSubtitle(DateFormat.format(format, patchDate).toString());
                } catch (ParseException e) {
                    e.printStackTrace();
                    item.setSubtitle(Build.VERSION.SECURITY_PATCH);
                }
            } else {
                item.setUnavailableitem(
                    new UnavailableItem(UnavailableType.NOT_POSSIBLE_YET,"6.0"));
            }
        } catch (Exception e) {
            Log.w(TAG, "Null in getAndroidVersion");
        }
        return item;
    }

    private Item getPreviewSDKInt(){
        Item item =new Item("Preview SDK Number", ItemType.SOFTWARE);
        try {
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                int sdk = Build.VERSION.PREVIEW_SDK_INT;
                if(sdk==0)
                    item.setSubtitle("Non-Preview");
                else
                    item.setSubtitle("Preview " + Integer.toString(sdk));
            }
            else {
                item.setUnavailableitem(
                    new UnavailableItem(UnavailableType.NOT_POSSIBLE_YET,"6.0"));
            }
        } catch (Exception e) {
            Log.w(TAG, "Null in getPreviewSDKInt");
        }
        return item;
    }

    private Item getDeviceBuildVersion() {
        //Get Moto specific build version if available
        Item item = new Item("Build Version", ItemType.SOFTWARE);
        SystemProperty sp = new SystemProperty(context);
        item.setSubtitle(sp.get("ro.build.version.full")==null|| sp.get("ro.build.version.full").equals("") ? Build.DISPLAY : sp.get("ro.build.version.full"));
        return item;
    }

    private Item getBuildBaseband() {
        Item item = new Item("Build Baseband", ItemType.SOFTWARE);
        item.setSubtitle(Build.getRadioVersion());
        return item;
    }

    private Item getBuildKernel() {
        Item item = new Item("Kernel Version", ItemType.SOFTWARE);
        item.setSubtitle(System.getProperty("os.version"));
        return item;
    }

    private Item getBuildDate() {
        Item item = new Item("Build Date", ItemType.SOFTWARE);
        item.setSubtitle(SimpleDateFormat.getInstance().format(new Date(Build.TIME)));
        return item;
    }

    private Item getBuildNumber() {
        Item item = new Item("Build Number", ItemType.SOFTWARE);
        item.setSubtitle(Build.ID);
        return item;
    }

    private Item getBuildBoard() {
        Item item = new Item("Build Board", ItemType.SOFTWARE);
        item.setSubtitle(Build.BOARD);
        return item;
    }

    private Item getBuildBootloader() {
        Item item = new Item("Build Bootloader", ItemType.SOFTWARE);
        item.setSubtitle(Build.BOOTLOADER);
        return item;
    }

    private Item getBuildBrand() {
        Item item = new Item("Build Brand", ItemType.SOFTWARE);
        item.setSubtitle(Build.BRAND);
        return item;
    }

    private Item getBuildDevice() {
        Item item = new Item("Build Brand", ItemType.SOFTWARE);
        item.setSubtitle(Build.DEVICE);
        return item;
    }

    private Item getBuildDisplay() {
        Item item = new Item("Build Display", ItemType.SOFTWARE);
        item.setSubtitle(Build.DISPLAY);
        return item;
    }

    private Item getBuildFingerprint() {
        Item item = new Item("Build Fingerprint", ItemType.SOFTWARE);
        item.setSubtitle(Build.FINGERPRINT);
        return item;
    }

    private Item getBuildHardware() {
        Item item = new Item("Build Hardware", ItemType.SOFTWARE);
        item.setSubtitle(Build.HARDWARE);
        return item;
    }

    private Item getBuildHost() {
        Item item = new Item("Build Host", ItemType.SOFTWARE);
        item.setSubtitle(Build.HOST);
        return item;
    }

    private Item getBuildTags() {
        Item item = new Item("Build Tags", ItemType.SOFTWARE);
        item.setSubtitle(Build.TAGS);
        return item;
    }

    private Item getBuildType() {
        Item item = new Item("Build Type", ItemType.SOFTWARE);
        item.setSubtitle(Build.TYPE);
        return item;
    }

    private Item getBuildUser() {
        Item item = new Item("Build User", ItemType.SOFTWARE);
        item.setSubtitle(Build.USER);
        return item;
    }

    private Item getOpenGLVersion(){
        Item item = new Item("OpenGL Version", ItemType.SOFTWARE);
        try {
            ConfigurationInfo configurationInfo = ((ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE)).getDeviceConfigurationInfo();
            item.setSubtitle(configurationInfo.getGlEsVersion());
        } catch (Exception e) {
            Log.e(TAG, "Exception in getOpenGLVersion");
        }
        return item;
    }

    private Item getGooglePlayServicesVersion() {
        Item item = new Item("Google Play Services Version", ItemType.SOFTWARE);
        try {
            String n = context.getPackageManager().getPackageInfo("com.google.android.gms", 0).versionName;
            int v = context.getPackageManager().getPackageInfo("com.google.android.gms", 0).versionCode;
            item.setSubtitle(n+" ("+String.valueOf(v)+")");
        } catch (Exception e) {
            Log.e(TAG, "Exception in getGooglePlayServicesVersion");
        }
        return item;
    }

    private Item getGooglePlayServicesInstallDate() {
        Item item = new Item("Google Play Services Installed", ItemType.SOFTWARE);
        try {
            long t = context.getPackageManager().getPackageInfo("com.google.android.gms", 0).firstInstallTime;
            item.setSubtitle(DateFormat.getDateFormat(context).format(t));
        } catch (Exception e) {
            Log.e(TAG, "Exception in getGooglePlayServicesVersion");
        }
        return item;
    }

    private Item getGooglePlayServicesUpdatedDate() {
        Item item = new Item("Google Play Services Updated", ItemType.SOFTWARE);
        try {
            long t = context.getPackageManager().getPackageInfo("com.google.android.gms", 0).lastUpdateTime;
            item.setSubtitle(DateFormat.getDateFormat(context).format(t));
        } catch (Exception e) {
            Log.e(TAG, "Exception in getGooglePlayServicesVersion");
        }
        return item;
    }
}

File: app/src/main/java/com/cwlarson/deviceid/database/Software.kt
package com.cwlarson.deviceid.database

import android.content.Context
import android.os.Build
import android.text.format.DateFormat
import androidx.core.content.pm.PackageInfoCompat
import com.cwlarson.deviceid.databinding.Item
import com.cwlarson.deviceid.databinding.ItemType
import com.cwlarson.deviceid.databinding.UnavailableItem
import com.cwlarson.deviceid.databinding.UnavailableType
import com.cwlarson.deviceid.util.SystemProperty
import com.cwlarson.deviceid.util.activityManager
import com.cwlarson.deviceid.util.gmsPackageInfo
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import timber.log.Timber
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Use [Build.VERSION.SDK_INT] to get the version number of the release
 */
fun Int.sdkToVersion(): String {
    return when(this) {
        Build.VERSION_CODES.BASE -> "1.0"
        Build.VERSION_CODES.BASE_1_1 -> "1.1"
        Build.VERSION_CODES.CUPCAKE -> "1.5"
        Build.VERSION_CODES.DONUT -> "1.6"
        Build.VERSION_CODES.ECLAIR -> "2.0"
        Build.VERSION_CODES.ECLAIR_0_1 -> "2.0.1"
        Build.VERSION_CODES.ECLAIR_MR1 -> "2.1"
        Build.VERSION_CODES.FROYO -> "2.2"
        Build.VERSION_CODES.GINGERBREAD -> "2.3"
        Build.VERSION_CODES.GINGERBREAD_MR1 -> "2.3.3"
        Build.VERSION_CODES.HONEYCOMB -> "3.0"
        Build.VERSION_CODES.HONEYCOMB_MR1 -> "3.1"
        Build.VERSION_CODES.HONEYCOMB_MR2 -> "3.2"
        Build.VERSION_CODES.ICE_CREAM_SANDWICH -> "4.0.1"
        Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1 -> "4.0.3"
        Build.VERSION_CODES.JELLY_BEAN -> "4.1"
        Build.VERSION_CODES.JELLY_BEAN_MR1 -> "4.2"
        Build.VERSION_CODES.JELLY_BEAN_MR2 -> "4.3"
        Build.VERSION_CODES.KITKAT -> "4.4"
        Build.VERSION_CODES.KITKAT_WATCH -> "4.4W"
        Build.VERSION_CODES.LOLLIPOP -> "5.0"
        Build.VERSION_CODES.LOLLIPOP_MR1 -> "5.1"
        Build.VERSION_CODES.M -> "6.0"
        Build.VERSION_CODES.N -> "7.0"
        Build.VERSION_CODES.N_MR1 -> "7.1"
        Build.VERSION_CODES.O -> "8.0"
        Build.VERSION_CODES.O_MR1 -> "8.1"
        Build.VERSION_CODES.P -> "9.0"
        Build.VERSION_CODES.Q -> "10.0"
        else -> "UNKNOWN"
    }
}

/**
 * Use [Build.VERSION.SDK_INT] to get the version name of the release
 * NOTE: No longer used after Android Pie
 */
private fun Int.getCodename(): String =
        when(this) {
            Build.VERSION_CODES.CUPCAKE -> "Cupcake"
            Build.VERSION_CODES.DONUT -> "Donut"
            Build.VERSION_CODES.ECLAIR, Build.VERSION_CODES.ECLAIR_0_1, Build.VERSION_CODES.ECLAIR_MR1 -> "Eclair"
            Build.VERSION_CODES.FROYO -> "Froyo"
            Build.VERSION_CODES.GINGERBREAD, Build.VERSION_CODES.GINGERBREAD_MR1 -> "Gingerbread"
            Build.VERSION_CODES.HONEYCOMB, Build.VERSION_CODES.HONEYCOMB_MR1, Build.VERSION_CODES.HONEYCOMB_MR2 -> "Honeycomb"
            Build.VERSION_CODES.ICE_CREAM_SANDWICH, Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1 -> "Ice Cream Sandwich"
            Build.VERSION_CODES.JELLY_BEAN, Build.VERSION_CODES.JELLY_BEAN_MR1, Build.VERSION_CODES.JELLY_BEAN_MR2 -> "Jelly Bean"
            Build.VERSION_CODES.KITKAT -> "KitKat"
            Build.VERSION_CODES.KITKAT_WATCH -> "KitKat Watch"
            Build.VERSION_CODES.LOLLIPOP, Build.VERSION_CODES.LOLLIPOP_MR1 -> "Lollipop"
            Build.VERSION_CODES.M -> "Marshmallow"
            Build.VERSION_CODES.N, Build.VERSION_CODES.N_MR1 -> "Nougat"
            Build.VERSION_CODES.O, Build.VERSION_CODES.O_MR1 -> "Oreo"
            Build.VERSION_CODES.P -> "Pie"
            else -> ""
        }

class Software(private val context: Context, db: AppDatabase, scope: CoroutineScope) {
    init {
        scope.launch {
            db.addItems(context, androidVersion(), patchLevel(), previewSDKInt(), deviceBuildVersion(),
                    buildBaseband(), buildKernel(), buildDate(), buildNumber(), buildBoard(),
                    buildBootloader(), buildBrand(), buildDevice(), buildDisplay(), buildFingerprint(),
                    buildHardware(), buildHost(), buildTags(), buildType(), buildUser(), openGLVersion(),
                    googlePlayServicesVersion(), googlePlayServicesInstallDate(),
                    googlePlayServicesUpdatedDate())
        }
    }

    private fun androidVersion() = Item("Android Version", ItemType.SOFTWARE).apply {
        try {
            subtitle = "${Build.VERSION.RELEASE} (API level ${Build.VERSION.SDK_INT}) ${Build.VERSION.SDK_INT.getCodename()}".trim()
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun patchLevel() = Item("Security Patch Level", ItemType.SOFTWARE).apply {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                subtitle = try {
                    val template = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
                    val patchDate = template.parse(Build.VERSION.SECURITY_PATCH)
                    val format = DateFormat.getBestDateTimePattern(Locale.getDefault(), "dMMMMyyyy")
                    DateFormat.format(format, patchDate).toString()
                } catch (e: ParseException) {
                    e.printStackTrace()
                    Build.VERSION.SECURITY_PATCH
                }
            } else {
                unavailableItem = UnavailableItem(UnavailableType.NOT_POSSIBLE_YET,
                        Build.VERSION_CODES.M.sdkToVersion())
            }
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun previewSDKInt() = Item("Preview SDK Number", ItemType.SOFTWARE).apply {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                subtitle = Build.VERSION.PREVIEW_SDK_INT.run {
                    if(this == 0) "Non-Preview" else "Preview $this"
                }
            } else {
                unavailableItem = UnavailableItem(UnavailableType.NOT_POSSIBLE_YET,
                        Build.VERSION_CODES.N.sdkToVersion())
            }
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    //Get Moto specific build version if available
    private fun deviceBuildVersion() = Item("Build Version", ItemType.SOFTWARE).apply {
        try {
            val sp = SystemProperty(context)
            subtitle = if (sp["ro.build.version.full"] == null || sp["ro.build.version.full"] == "") Build.DISPLAY else sp["ro.build.version.full"]
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildBaseband() = Item("Build Baseband", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.getRadioVersion()
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildKernel() = Item("Kernel Version", ItemType.SOFTWARE).apply {
        try {
            subtitle = System.getProperty("os.version")
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildDate() = Item("Build Date", ItemType.SOFTWARE).apply {
        try {
            subtitle = SimpleDateFormat.getInstance().format(Date(Build.TIME))
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildNumber() = Item("Build Number", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.ID
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildBoard() = Item("Build Board", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.BOARD
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildBootloader() = Item("Build Bootloader", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.BOOTLOADER
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildBrand() = Item("Build Brand", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.BRAND
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildDevice() = Item("Build Device", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.DEVICE
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildDisplay() = Item("Build Display", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.DISPLAY
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildFingerprint() = Item("Build Fingerprint", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.FINGERPRINT
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildHardware() = Item("Build Hardware", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.HARDWARE
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildHost() = Item("Build Host", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.HOST
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildTags() = Item("Build Tags", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.TAGS
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildType() = Item("Build Type", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.TYPE
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun buildUser() = Item("Build User", ItemType.SOFTWARE).apply {
        try {
            subtitle = Build.USER
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun openGLVersion() = Item("OpenGL Version", ItemType.SOFTWARE).apply {
        try {
            subtitle = context.activityManager.deviceConfigurationInfo.glEsVersion
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun googlePlayServicesVersion() = Item("Google Play Services Version", ItemType.SOFTWARE).apply {
        try {
            val pi = context.gmsPackageInfo
            val v = PackageInfoCompat.getLongVersionCode(pi)
            subtitle = "${pi.versionName} ($v)"
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun googlePlayServicesInstallDate() = Item("Google Play Services Installed", ItemType.SOFTWARE).apply {
        try {
            subtitle = DateFormat.getDateFormat(context).format(context.gmsPackageInfo.firstInstallTime)
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun googlePlayServicesUpdatedDate() = Item("Google Play Services Updated", ItemType.SOFTWARE).apply {
        try {
            subtitle = DateFormat.getDateFormat(context).format(context.gmsPackageInfo.lastUpdateTime)
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }
}
